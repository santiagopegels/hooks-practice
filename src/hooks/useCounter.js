import { useState } from "react"

export const useCounter = (initialState = 10) => {
    const [state, setState] = useState(initialState)

    const reset = () => {
        setState(initialState)
    }

    const increment = () => {
        setState(state + 1)
    }

    const decrement = (factor=1) => {
        setState(state - factor)
    }

    return {
        state,
        increment,
        decrement,
        reset
    }
}
