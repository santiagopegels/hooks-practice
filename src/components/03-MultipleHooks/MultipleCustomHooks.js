import React from 'react'
import { useFetch } from '../../hooks/useFetch'
import { useCounter } from '../../hooks/useCounter'

export const MultipleCustomHooks = () => {

    const { state: counter, increment } = useCounter(1)

    const { loading, data } = useFetch(`https://breakingbadapi.com/api/quotes/${counter}`)

    const { author, quote } = !!data && data[0]

    return (
        <div>
            <h1>Breaking Bad Quotes</h1>
            <hr></hr>

            {
                loading
                    ?
                    (
                        <div className="alert alert-success text-center">
                            Loading...
                        </div>
                    )
                    :
                    (
                        <blockquote className="blockquote text-end">
                            <p className="mb-3"> {quote} </p>
                            <footer className="blockquote-footer">{author}</footer>

                        </blockquote>
                    )
            }

            {
                !loading &&
                <button
                    className="btn btn-primary"
                    onClick={() => increment(1)}
                >
                    Next Quote
            </button>
            }
        </div>
    )
}
