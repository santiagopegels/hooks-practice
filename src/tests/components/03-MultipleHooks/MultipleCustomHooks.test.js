import React from 'react'
import { shallow } from 'enzyme'
import {MultipleCustomHooks} from '../../../components/03-MultipleHooks/MultipleCustomHooks'
import { useFetch } from '../../../hooks/useFetch'
import { useCounter } from '../../../hooks/useCounter'


jest.mock('../../../hooks/useFetch')
jest.mock('../../../hooks/useCounter')

beforeEach( () => {
    useCounter.mockReturnValue({
        state: 10,
        increment: () => {}
    })
})

describe('<MultipleCustomHooks /> tests', () => {

test('should show correctly', () => {

    useFetch.mockReturnValue({
        data: null, 
        loading: true,
        error:null
    })

    const wrapper = shallow(<MultipleCustomHooks />)
    expect(wrapper).toMatchSnapshot()
})

test('should show the information', () => {
    
    const data = [{
        author: 'Test',
        quote: 'Test Text'
    }]

    useFetch.mockReturnValue({
        data, 
        loading: false,
        error:null
    })

    const wrapper = shallow(<MultipleCustomHooks />)

    expect(wrapper.find('.alert').exists()).toBe(false)
    expect(wrapper.find('.mb-3').text().trim()).toBe(data[0].quote)
    expect(wrapper.find('footer').text().trim()).toBe(data[0].author)

})

})
