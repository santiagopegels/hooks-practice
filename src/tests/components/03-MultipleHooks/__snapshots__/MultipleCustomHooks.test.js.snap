// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultipleCustomHooks /> tests should show correctly 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-success text-center"
  >
    Loading...
  </div>
</div>
`;
